src/utils/string_utils.hpp

```diff
@@ -27,62 +27,95 @@ namespace nmodl {
 namespace stringutils {
 
 /**
- * @addtogroup utils
- * @{
+ * \addtogroup utils
+ * \{
  */
 
 /// text alignment when printing in the tabular form
 enum class text_alignment { left, right, center };
 
-/// Trim from start
-static inline std::string& ltrim(std::string& s) {
-    s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](int c) { return !std::isspace(c); }));
-    return s;
+/**
+ * \param text the string to manipulate
+ * \return a copy of the given string with leading ASCII space characters removed
+ */
+[[nodiscard]] static inline std::string ltrim(std::string text) {
+    text.erase(text.begin(),
+               std::find_if(text.begin(), text.end(), [](int c) { return !std::isspace(c); }));
+    return text;
 }
 
-/// Trim from end
-static inline std::string& rtrim(std::string& s) {
-    s.erase(std::find_if(s.rbegin(), s.rend(), [](int c) { return !std::isspace(c); }).base(),
-            s.end());
-    return s;
+/**
+ * \param text the string to manipulate
+ * \return a copy of the given string with trailing characters removed.
+ */
+[[nodiscard]] static inline std::string rtrim(std::string text) {
+    text.erase(
+        std::find_if(text.rbegin(), text.rend(), [](int c) { return !std::isspace(c); }).base(),
+        text.end());
+    return text;
 }
 
-/// Trim from both ends
-static inline std::string& trim(std::string& s) {
-    return ltrim(rtrim(s));
+/**
+ *
+ * \param text the string to manipulate
+ * \return a copy of the given string with both leading and trailing ASCII space characters removed
+ */
+[[nodiscard]] static inline std::string trim(const std::string& text) {
+    return ltrim(rtrim(text));
 }
 
-static inline void remove_character(std::string& str, const char c) {
-    str.erase(std::remove(str.begin(), str.end(), c), str.end());
+/**
+ * Remove all occurrences of a given character in a text
+ * \param text the string to manipulate
+ * \param c the character to remove
+ * @return a copy the modified text
+ */
+[[nodiscard]] static inline std::string remove_character(std::string text, const char c) {
+    text.erase(std::remove(text.begin(), text.end(), c), text.end());
+    return text;
 }
 
-/// Remove leading newline for the string read by grammar
-static inline std::string& trim_newline(std::string& s) {
-    remove_character(s, '\n');
-    return s;
+/**
+ *
+ * \param text the string to manipulate
+ * \return a copy of the given string with all occurrences of the ASCII newline character removed
+ */
+[[nodiscard]] static inline std::string trim_newline(const std::string& text) {
```

Comment:
Most functions take a `std::string` this one takes a const reference, why?

Reply:
This function is just a pass-through function `remove_character`, no copy required.

Reply:
But `remove_character` accepts an `std::string`. If you pass it a const-ref to an `std::string` this will force a copy. If `trim_newline` accepts by value and moves, then you always pay for one move, sometimes a copy, but the copy can be avoided if you move into `trim_newline`.

Reply:
To make sure I understand, your point is that a copy will be made anyhow. So let's keep the API consistent with functions accepting a `std::string`, and make the "efficient" move an implementation detail. Something like below @1uc :

```diff
diff --git a/src/utils/string_utils.hpp b/src/utils/string_utils.hpp
index 2a48fc24..c11ab18b 100644
--- a/src/utils/string_utils.hpp
+++ b/src/utils/string_utils.hpp
@@ -60,8 +60,8 @@ enum class text_alignment { left, right, center };
  * \param text the string to manipulate
  * \return a copy of the given string with both leading and trailing ASCII space characters removed
  */
-[[nodiscard]] static inline std::string trim(const std::string& text) {
-    return ltrim(rtrim(text));
+[[nodiscard]] static inline std::string trim(std::string text) {
+    return ltrim(rtrim(std::move(text)));
 }

 /**
@@ -80,8 +80,8 @@ enum class text_alignment { left, right, center };
  * \param text the string to manipulate
  * \return a copy of the given string with all occurrences of the ASCII newline character removed
  */
-[[nodiscard]] static inline std::string trim_newline(const std::string& text) {
-    return remove_character(text, '\n');
+[[nodiscard]] static inline std::string trim_newline(std::string text) {
+    return remove_character(std::move(text), '\n');
 }

 /**
```

Reply:
Nice suggestion. added in [f3f3a70](https://github.com/BlueBrain/nmodl/pull/1071/commits/f3f3a700ce5c9efbe594338cf84e74ab50e796f9)